File: src/SSHServer/Packets/KexInit.cs
using SSHServer.Ciphers;
using SSHServer.Compressions;
using SSHServer.HostKeyAlgorithms;
using SSHServer.KexAlgorithms;
using SSHServer.MACAlgorithms;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Threading.Tasks;

namespace SSHServer.Packets
{
    public class KexInit : Packet
    {
        public override PacketType PacketType
        {
            get
            {
                return PacketType.SSH_MSG_KEXINIT;
            }
        }

        public byte[] Cookie { get; set; } = new byte[16];
        public List<string> KexAlgorithms { get; private set; } = new List<string>();
        public List<string> ServerHostKeyAlgorithms { get; private set; } = new List<string>();
        public List<string> EncryptionAlgorithmsClientToServer { get; private set; } = new List<string>();
        public List<string> EncryptionAlgorithmsServerToClient { get; private set; } = new List<string>();
        public List<string> MacAlgorithmsClientToServer { get; private set; } = new List<string>();
        public List<string> MacAlgorithmsServerToClient { get; private set; } = new List<string>();
        public List<string> CompressionAlgorithmsClientToServer { get; private set; } = new List<string>();
        public List<string> CompressionAlgorithmsServerToClient { get; private set; } = new List<string>();
        public List<string> LanguagesClientToServer { get; private set; } = new List<string>();
        public List<string> LanguagesServerToClient { get; private set; } = new List<string>();
        public bool FirstKexPacketFollows { get; set; }

        public KexInit()
        {
            RandomNumberGenerator.Create().GetBytes(Cookie);
        }

        protected override void InternalGetBytes(ByteWriter writer)
        {
            writer.WriteRawBytes(Cookie);
            writer.WriteStringList(KexAlgorithms);
            writer.WriteStringList(ServerHostKeyAlgorithms);
            writer.WriteStringList(EncryptionAlgorithmsClientToServer);
            writer.WriteStringList(EncryptionAlgorithmsServerToClient);
            writer.WriteStringList(MacAlgorithmsClientToServer);
            writer.WriteStringList(MacAlgorithmsServerToClient);
            writer.WriteStringList(CompressionAlgorithmsClientToServer);
            writer.WriteStringList(CompressionAlgorithmsServerToClient);
            writer.WriteStringList(LanguagesClientToServer);
            writer.WriteStringList(LanguagesServerToClient);
            writer.WriteByte(FirstKexPacketFollows ? (byte)0x01 : (byte)0x00);
            writer.WriteUInt32(0);
        }

        public override void Load(ByteReader reader)
        {
            Cookie = reader.GetBytes(16);
            KexAlgorithms = reader.GetNameList();
            ServerHostKeyAlgorithms = reader.GetNameList();
            EncryptionAlgorithmsClientToServer = reader.GetNameList();
            EncryptionAlgorithmsServerToClient = reader.GetNameList();
            MacAlgorithmsClientToServer = reader.GetNameList();
            MacAlgorithmsServerToClient = reader.GetNameList();
            CompressionAlgorithmsClientToServer = reader.GetNameList();
            CompressionAlgorithmsServerToClient = reader.GetNameList();
            LanguagesClientToServer = reader.GetNameList();
            LanguagesServerToClient = reader.GetNameList();
            FirstKexPacketFollows = reader.GetBoolean();
            /*
              uint32       0 (reserved for future extension)
            */
            uint reserved = reader.GetUInt32();
        }

        public IKexAlgorithm PickKexAlgorithm()
        {
            foreach (string algo in this.KexAlgorithms)
            {
                IKexAlgorithm selectedAlgo = Server.GetType<IKexAlgorithm>(Server.SupportedKexAlgorithms, algo);
                if (selectedAlgo != null)
                {
                    return selectedAlgo;
                }
            }

            // If no algorithm satisfying all these conditions can be found, the
            // connection fails, and both sides MUST disconnect.
            throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "Could not find a shared Kex Algorithm");
        }

        public IHostKeyAlgorithm PickHostKeyAlgorithm()
        {
            foreach (string algo in this.ServerHostKeyAlgorithms)
            {
                IHostKeyAlgorithm selectedAlgo = Server.GetType<IHostKeyAlgorithm>(Server.SupportedHostKeyAlgorithms, algo);
                if (selectedAlgo != null)
                {
                    return selectedAlgo;
                }
            }

            // If no algorithm satisfying all these conditions can be found, the
            // connection fails, and both sides MUST disconnect.
            throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "Could not find a shared Host Key Algorithm");
        }

        public ICipher PickCipherClientToServer()
        {
            foreach (string algo in this.EncryptionAlgorithmsClientToServer)
            {
                ICipher selectedCipher = Server.GetType<ICipher>(Server.SupportedCiphers, algo);
                if (selectedCipher != null)
                {
                    return selectedCipher;
                }
            }

            // If no algorithm satisfying all these conditions can be found, the
            // connection fails, and both sides MUST disconnect.
            throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "Could not find a shared Client-To-Server Cipher Algorithm");
        }

        public ICipher PickCipherServerToClient()
        {
            foreach (string algo in this.EncryptionAlgorithmsServerToClient)
            {
                ICipher selectedCipher = Server.GetType<ICipher>(Server.SupportedCiphers, algo);
                if (selectedCipher != null)
                {
                    return selectedCipher;
                }
            }

            // If no algorithm satisfying all these conditions can be found, the
            // connection fails, and both sides MUST disconnect.
            throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "Could not find a shared Server-To-Client Cipher Algorithm");
        }

        public IMACAlgorithm PickMACAlgorithmClientToServer()
        {
            foreach (string algo in this.MacAlgorithmsClientToServer)
            {
                IMACAlgorithm selectedAlgo = Server.GetType<IMACAlgorithm>(Server.SupportedMACAlgorithms, algo);
                if (selectedAlgo != null)
                {
                    return selectedAlgo;
                }
            }

            // If no algorithm satisfying all these conditions can be found, the
            // connection fails, and both sides MUST disconnect.
            throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "Could not find a shared Client-To-Server MAC Algorithm");
        }

        public IMACAlgorithm PickMACAlgorithmServerToClient()
        {
            foreach (string algo in this.MacAlgorithmsServerToClient)
            {
                IMACAlgorithm selectedAlgo = Server.GetType<IMACAlgorithm>(Server.SupportedMACAlgorithms, algo);
                if (selectedAlgo != null)
                {
                    return selectedAlgo;
                }
            }

            // If no algorithm satisfying all these conditions can be found, the
            // connection fails, and both sides MUST disconnect.
            throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "Could not find a shared Server-To-Client MAC Algorithm");
        }

        public ICompression PickCompressionAlgorithmClientToServer()
        {
            foreach (string algo in this.CompressionAlgorithmsClientToServer)
            {
                ICompression selectedAlgo = Server.GetType<ICompression>(Server.SupportedCompressions, algo);
                if (selectedAlgo != null)
                {
                    return selectedAlgo;
                }
            }

            // If no algorithm satisfying all these conditions can be found, the
            // connection fails, and both sides MUST disconnect.
            throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "Could not find a shared Client-To-Server Compression Algorithm");
        }

        public ICompression PickCompressionAlgorithmServerToClient()
        {
            foreach (string algo in this.CompressionAlgorithmsServerToClient)
            {
                ICompression selectedAlgo = Server.GetType<ICompression>(Server.SupportedCompressions, algo);
                if (selectedAlgo != null)
                {
                    return selectedAlgo;
                }
            }

            // If no algorithm satisfying all these conditions can be found, the
            // connection fails, and both sides MUST disconnect.
            throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "Could not find a shared Server-To-Client Compresion Algorithm");
        }
    }
}

File: SSHServerNodeJS/SSHServerNodeJS/Packets/KexInit.ts
import { Packet } from "./Packet";
import { Server } from "../Server";
import { PacketType } from "./PacketType";
import { ByteReader } from "../ByteReader";
import { ByteWriter } from "../ByteWriter";

import { IKexAlgorithm } from "../KexAlgorithms/IKexAlgorithm";
import { DiffieHellmanGroup14SHA1 } from "../KexAlgorithms/DiffieHellmanGroup14SHA1";

import { IHostKeyAlgorithm } from "../HostKeyAlgorithms/IHostKeyAlgorithm";
import { SSHRSA } from "../HostKeyAlgorithms/SSHRSA";

import { ICipher } from "../Ciphers/ICipher";
import { TripleDESCBC } from "../Ciphers/TripleDESCBC";

import { IMACAlgorithm } from "../MACAlgorithms/IMACAlgorithm";
import { HMACSHA1 } from "../MACAlgorithms/HMACSHA1";

import { ICompression } from "../Compressions/ICompression";
import { NoCompression } from "../Compressions/NoCompression";

import * as Exceptions from "../SSHServerException";

import crypto = require('crypto');

export class KexInit extends Packet {
    public getPacketType(): PacketType {
        return PacketType.SSH_MSG_KEXINIT;
    }

    public cookie: Buffer = crypto.randomBytes(16);
    public kexAlgorithms: Array<string> = new Array<string>();
    public serverHostKeyAlgorithms : Array<string> = new Array<string>();
    public encryptionAlgorithmsClientToServer : Array<string> = new Array<string>();
    public encryptionAlgorithmsServerToClient : Array<string> = new Array<string>();
    public macAlgorithmsClientToServer : Array<string> = new Array<string>();
    public macAlgorithmsServerToClient : Array<string> = new Array<string>();
    public compressionAlgorithmsClientToServer : Array<string> = new Array<string>();
    public compressionAlgorithmsServerToClient : Array<string> = new Array<string>();
    public languagesClientToServer : Array<string> = new Array<string>();
    public languagesServerToClient: Array<string> = new Array<string>();
    public firstKexPacketFollows: boolean = false;

    protected internalGetBytes(writer: ByteWriter) {
        writer.writeRawBytes(this.cookie);
        writer.writeStringList(this.kexAlgorithms);
        writer.writeStringList(this.serverHostKeyAlgorithms);
        writer.writeStringList(this.encryptionAlgorithmsClientToServer);
        writer.writeStringList(this.encryptionAlgorithmsServerToClient);
        writer.writeStringList(this.macAlgorithmsClientToServer);
        writer.writeStringList(this.macAlgorithmsServerToClient);
        writer.writeStringList(this.compressionAlgorithmsClientToServer);
        writer.writeStringList(this.compressionAlgorithmsServerToClient);
        writer.writeStringList(this.languagesClientToServer);
        writer.writeStringList(this.languagesServerToClient);
        writer.writeByte(this.firstKexPacketFollows ? 0x01 : 0x00);
        writer.writeUInt32(0);
    }

    public load(reader: ByteReader) {
        this.cookie = reader.getBytes(16);
        this.kexAlgorithms = reader.getNameList();
        this.serverHostKeyAlgorithms = reader.getNameList();
        this.encryptionAlgorithmsClientToServer = reader.getNameList();
        this.encryptionAlgorithmsServerToClient = reader.getNameList();
        this.macAlgorithmsClientToServer = reader.getNameList();
        this.macAlgorithmsServerToClient = reader.getNameList();
        this.compressionAlgorithmsClientToServer = reader.getNameList();
        this.compressionAlgorithmsServerToClient = reader.getNameList();
        this.languagesClientToServer = reader.getNameList();
        this.languagesServerToClient = reader.getNameList();
        this.firstKexPacketFollows = reader.getBoolean();

        // uint32       0 (reserved for future extension)
        reader.getUInt32();
    }

    public pickKexAlgorithm(): IKexAlgorithm {
        for (let algo of this.kexAlgorithms) {
            if (Server.SupportedKexAlgorithms.find(
                (value: string, index: number, obj: Array<string>): boolean => {
                    return (value === algo);
                })) {

                switch (algo) {
                    case "diffie-hellman-group14-sha1":
                        return new DiffieHellmanGroup14SHA1();
                }
            }
        }

        // If no algorithm satisfying all these conditions can be found, the
        // connection fails, and both sides MUST disconnect.
        throw new Exceptions.SSHServerException(
            Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
            "Could not find a shared Kex Algorithm");
    }

    public pickHostKeyAlgorithm(): IHostKeyAlgorithm {
        for (let algo of this.serverHostKeyAlgorithms) {
            if (Server.SupportedHostKeyAlgorithms.find(
                (value: string, index: number, obj: Array<string>): boolean => {
                    return (value === algo);
                })) {

                switch (algo) {
                    case "ssh-rsa":
                        return new SSHRSA();
                }
            }
        }

        // If no algorithm satisfying all these conditions can be found, the
        // connection fails, and both sides MUST disconnect.
        throw new Exceptions.SSHServerException(
            Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
            "Could not find a shared Host Key Algorithm");
    }

    public pickCipherClientToServer(): ICipher {
        for (let algo of this.encryptionAlgorithmsClientToServer) {
            if (Server.SupportedCiphers.find(
                (value: string, index: number, obj: Array<string>): boolean => {
                    return (value === algo);
                })) {

                switch (algo) {
                    case "3des-cbc":
                        return new TripleDESCBC();
                }
            }
        }

        // If no algorithm satisfying all these conditions can be found, the
        // connection fails, and both sides MUST disconnect.
        throw new Exceptions.SSHServerException(
            Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
            "Could not find a shared Client-To-Server Cipher Algorithm");
    }

    public pickCipherServerToClient(): ICipher {
        for (let algo of this.encryptionAlgorithmsServerToClient) {
            if (Server.SupportedCiphers.find(
                (value: string, index: number, obj: Array<string>): boolean => {
                    return (value === algo);
                })) {

                switch (algo) {
                    case "3des-cbc":
                        return new TripleDESCBC();
                }
            }
        }

        // If no algorithm satisfying all these conditions can be found, the
        // connection fails, and both sides MUST disconnect.
        throw new Exceptions.SSHServerException(
            Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
            "Could not find a shared Server-To-Client Cipher Algorithm");
    }

    public pickMACAlgorithmClientToServer(): IMACAlgorithm {
        for (let algo of this.macAlgorithmsClientToServer) {
            if (Server.SupportedMACAlgorithms.find(
                (value: string, index: number, obj: Array<string>): boolean => {
                    return (value === algo);
                })) {

                switch (algo) {
                    case "hmac-sha1":
                        return new HMACSHA1();
                }
            }
        }

        // If no algorithm satisfying all these conditions can be found, the
        // connection fails, and both sides MUST disconnect.
        throw new Exceptions.SSHServerException(
            Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
            "Could not find a shared Client-To-Server MAC Algorithm");
    }

    public pickMACAlgorithmServerToClient(): IMACAlgorithm {
        for (let algo of this.macAlgorithmsServerToClient) {
            if (Server.SupportedMACAlgorithms.find(
                (value: string, index: number, obj: Array<string>): boolean => {
                    return (value === algo);
                })) {

                switch (algo) {
                    case "hmac-sha1":
                        return new HMACSHA1();
                }
            }
        }

        // If no algorithm satisfying all these conditions can be found, the
        // connection fails, and both sides MUST disconnect.
        throw new Exceptions.SSHServerException(
            Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
            "Could not find a shared Server-To-Client MAC Algorithm");
    }

    public pickCompressionAlgorithmClientToServer(): ICompression {
        for (let algo of this.compressionAlgorithmsClientToServer) {
            if (Server.SupportedCompressions.find(
                (value: string, index: number, obj: Array<string>): boolean => {
                    return (value === algo);
                })) {

                switch (algo) {
                    case "none":
                        return new NoCompression();
                }
            }
        }

        // If no algorithm satisfying all these conditions can be found, the
        // connection fails, and both sides MUST disconnect.
        throw new Exceptions.SSHServerException(
            Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
            "Could not find a shared Client-To-Server Compression Algorithm");
    }

    public pickCompressionAlgorithmServerToClient(): ICompression {
        for (let algo of this.compressionAlgorithmsServerToClient) {
            if (Server.SupportedCompressions.find(
                (value: string, index: number, obj: Array<string>): boolean => {
                    return (value === algo);
                })) {

                switch (algo) {
                    case "none":
                        return new NoCompression();
                }
            }
        }

        // If no algorithm satisfying all these conditions can be found, the
        // connection fails, and both sides MUST disconnect.
        throw new Exceptions.SSHServerException(
            Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
            "Could not find a shared Server-To-Client Compresion Algorithm");
    }
}
